app/frontend/wallet/account-manager.ts

```diff
@@ -46,29 +49,66 @@ const AccountManager = ({
   }
 
   async function discoverAccounts() {
+    if (
+      maxAccountIndex === accounts.length - 1 || accounts.length > 0
+        ? !(await accounts[accounts.length - 1].isAccountUsed())
+        : false
+    ) {
+      return accounts
+    }
+
     const isBulkExportSupported = cryptoProvider.isFeatureSupported(
       CryptoProviderFeature.BULK_EXPORT
     )
     const shouldExplore =
       config.shouldExportPubKeyBulk && config.isShelleyCompatible && isBulkExportSupported
-    async function _discoverNextAccount(accountIndex: number) {
-      const newAccount = accounts[accountIndex] || discoverNextAccount()
-      const isAccountUsed = await newAccount.isAccountUsed()
-      if (accountIndex === accounts.length) await addNextAccount(newAccount)
-
-      return (
-        shouldExplore &&
-        isAccountUsed &&
-        accountIndex < maxAccountIndex &&
-        (await _discoverNextAccount(accountIndex + 1))
+
+    if (!shouldExplore) {
+      await exploreNextAccount()
+      return accounts
+    }
+
+    for (const [accountIndexStart, accountIndexEnd] of makeBulkAccountIndexIterator()) {
+      const _accountIndexStart = Math.max(accountIndexStart, accounts.length)
+      const _accountIndexEnd = Math.min(accountIndexEnd, maxAccountIndex)
+
+      if (_accountIndexStart > _accountIndexEnd) {
+        continue
+      }
+
+      const newAccountBatch = _.range(_accountIndexStart, _accountIndexEnd + 1).map(
+        (accountIndex) => accounts[accountIndex] || discoverAccount(accountIndex)
       )
+
+      // Needs to be sequential because HW wallets can't process parallel calls
+      // xpubs keys are used in the next step when deciding if accounts are used
+      for (const account of newAccountBatch) {
+        await account.ensureXpubIsExported()
+      }
+
+      const areAccountsUsed = await Promise.all(
+        newAccountBatch.map((newAccount) => newAccount.isAccountUsed())
+      )
+
+      const firstUnusedIndex = areAccountsUsed.findIndex((isUsed) => isUsed === false)
+      const shouldContinueExploring = firstUnusedIndex === -1
```

Comment:
The `areAccountsUsed` intermediate structure seems a bit confusing to me, I'd find it easier to reason about a `firstUnusedNewAccountIndex` ("new account" to not confuse it with an account index in the derivation path)

```suggestion
      const firstUnusedNewAccountIndex = (await Promise.all(
        newAccountBatch.map((newAccount) => newAccount.isAccountUsed())
      )).findIndex((isUsed) => isUsed === false)

      const shouldContinueExploring = firstUnusedIndex === -1
```

Reply:
Refactored.